src/gui/search/searchjobwidget.cpp

```diff
@@ -168,6 +169,14 @@ SearchJobWidget::~SearchJobWidget()
 
 void SearchJobWidget::onItemDoubleClicked(const QModelIndex &index)
 {
+    if (!BitTorrent::Session::instance()->isRestored())
+    {
+        QMessageBox::warning(this, tr("BitTorrent Session isn't restored")
```

Comment:
(Globally) would it be possible to just disable the GUI widgets/elements instead of needing this message?
I think it is too intrusive... the message doesn't closely related to the action performed and the failing is unexpected for users.

IMO this is quite a big change... now is adding/enforcing a new *step* (restoring torrents) when starting up and when processing this step all session related operations are halted.
Maybe instead of enable/disable widgets here and there maybe present a new modal dialog on top of mainwindow which shows the restore progress (a progressbar with cancel button?), since the GUI doesn't have much functionality without session ready anyway...

Also what about webapi/webui users? How would they know when the session is ready?

Reply:
> now is adding/enforcing a new _step_ (restoring torrents) when starting up and when processing this step all session related operations are halted.

It's only about adding torrents. Most of the other operations are still available. And I highly doubt that earlier you would have been able to successfully add a torrent before resuming the session, given that the application was almost completely unresponsive during this period.
>Maybe instead of enable/disable widgets here and there maybe present a new modal dialog on top of mainwindow which shows the restore progress (a progressbar with cancel button?), since the GUI doesn't have much functionality without session ready anyway...

I really don't want to redo everything for now.  In addition, I believe that we will not understand how it would be better until it gets into production and we begin to receive real feedbacks.
>Also what about webapi/webui users? How would they know when the session is ready?

Please, just don't make me get into Web UI. Poking around there is beyond my strength. Honestly, it's easier for me to give it all up...

Reply:
I think words such as "BitTorrent" "session" and "restored" should not be used. Instead something simpler that can be understood by the average user, like "Torrents are still being loaded, please wait for the process to finish."

Reply:
>the message doesn't closely related to the action performed and the failing is unexpected for users.

Maybe just silently ignore "double click" until the session is restored?

Reply:
>Please, just don't make me get into Web UI. Poking around there is beyond my strength. Honestly, it's easier for me to give it all up...

I don't insist on webui. As for webapi, I'm not saying it is a hard requirement for this PR but it is something to think about (as to avoid redesigning the internals).

>Maybe just silently ignore "double click" until the session is restored?

I don't think users would like that.

>In fact, a UI that is unavailable during startup, displaying only the progress of execution, could get rid of most possible conflicts, and would speed up the startup itself.

(just providing ideas)
On top of the transferlist there could be a modal dialog like this (without the Close button):
![screenshot](https://docs.microsoft.com/en-us/windows/win32/uxguide/images/progress-bars-image4.png)
The transferlist won't be populated when this dialog is running.

To be honest, I don't really like disabling menu items/widgets due to session not ready so I'm finding alternative ways.

Reply:
> I don't really like disabling menu items/widgets due to session not ready

Yes, it looks quite cumbersome and poorly scalable.
So a modal dialog would be a better alternative. But I am not sure that this will work correctly in the case of starting hidden in the tray (without displaying the main window). I must try...
Other alternatives I'm thinking of:
1. Don't create any regular content of main window until session is restored and just display startup progress,
2. Don't create main window until session is restored and display startup progress in its own window.

Reply:
>Don't create any regular content of main window until session is restored and just display startup progress,

I'm would prefer this (leave the transferlist empty or disabled) and this way would be simpler to implement I presume?

Reply:
> > Don't create any regular content of main window until session is restored and just display startup progress,
> 
> I'm would prefer this (show the mainwindow but leave the transferlist empty or disabled) and this way would be simpler to implement I presume?

You don't seem to understand this option of mine that you commented on. I mean to create Main window that doesn't contain any regular controls, toolbox, menus etc. but only temporary controls needed to display startup progress. After session is restored the main window will expose its regular layout.
Otherwise we need either disable all regular controls or show modal dialog to make them unavailable.

Reply:
>I mean to create Main window that doesn't contain any regular controls, toolbox, menus etc. but only temporary controls needed to display startup progress. After session is restored the main window will expose its regular layout.

Then the mainwindow will have different layouts depending on the state of session, which is more work.

>The transferlist won't be populated when this dialog is running.

To clarify what I said before.
I meant on startup, the mainwindow is shown entirely (all menus, widgets) but the transferlist is empty. Next the session restore modal dialog is shown automatically (and closed automatically when done). While the session restore procedure runs, the transferlist is still empty, only when the procedure is finished then the transferlist will populate the contents.
Would this be possible? will there be downsides? I imagine this way would be easier than other options.

Reply:
>So a modal dialog would be a better alternative. But I am not sure that this will work correctly in the case of starting hidden in the tray (without displaying the main window). I must try...

(On Windows) currently dialogs (about, options, etc.) remains on screen when mainwindow is minimized to tray, so maybe just show the modal dialog anyway (at least for this first version).

Reply:
> > So a modal dialog would be a better alternative. But I am not sure that this will work correctly in the case of starting hidden in the tray (without displaying the main window). I must try...
> 
> (On Windows) currently dialogs (about, options, etc.) remains on screen when mainwindow is minimized to tray, so maybe just show the modal dialog anyway (at least for this first version, we can tweak it later).

I'm against windows showing when minimized to tray, as Qt sometimes crashes on monitor disconnect if it's drawing *any* windows (so I always make sure qbit is minimized to tray before I turn my screen off for the night)

Reply:
> (On Windows) currently dialogs (about, options, etc.) remains on screen when mainwindow is minimized to tray, so maybe just show the modal dialog anyway (at least for this first version, we can tweak it later).

I wouldn't consider it as acceptable solution. If users select to start qBittorrent minimized to tray they clearly don't want to see anything on the screen.


> Then the mainwindow will have different layouts depending on the state of session, which is more coding work.

Not very much. The creation of a regular layout could simply be placed in a separate method called when the session is restored.